<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="es" xml:lang="es"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Juan Gómez Romero">

<title>Clasificación con el conjunto de datos Iris</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="iris_files/libs/clipboard/clipboard.min.js"></script>
<script src="iris_files/libs/quarto-html/quarto.js"></script>
<script src="iris_files/libs/quarto-html/popper.min.js"></script>
<script src="iris_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="iris_files/libs/quarto-html/anchor.min.js"></script>
<link href="iris_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="iris_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="iris_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="iris_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="iris_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<link href="iris_files/libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet">
<script src="iris_files/libs/pagedtable-1.1/js/pagedtable.js"></script>


<link rel="stylesheet" href="styles.css">
</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Tabla de contenidos</h2>
   
  <ul>
  <li><a href="#lectura-de-datos" id="toc-lectura-de-datos" class="nav-link active" data-scroll-target="#lectura-de-datos"><span class="header-section-number">1</span> Lectura de datos</a></li>
  <li><a href="#análisis-exploratorio" id="toc-análisis-exploratorio" class="nav-link" data-scroll-target="#análisis-exploratorio"><span class="header-section-number">2</span> Análisis exploratorio</a>
  <ul class="collapse">
  <li><a href="#características" id="toc-características" class="nav-link" data-scroll-target="#características"><span class="header-section-number">2.1</span> Características</a></li>
  <li><a href="#histograma" id="toc-histograma" class="nav-link" data-scroll-target="#histograma"><span class="header-section-number">2.2</span> Histograma</a></li>
  <li><a href="#densidad" id="toc-densidad" class="nav-link" data-scroll-target="#densidad"><span class="header-section-number">2.3</span> Densidad</a></li>
  <li><a href="#dispersión" id="toc-dispersión" class="nav-link" data-scroll-target="#dispersión"><span class="header-section-number">2.4</span> Dispersión</a></li>
  </ul></li>
  <li><a href="#clasificación-con-k-nn" id="toc-clasificación-con-k-nn" class="nav-link" data-scroll-target="#clasificación-con-k-nn"><span class="header-section-number">3</span> Clasificación con k-NN</a>
  <ul class="collapse">
  <li><a href="#preprocesamiento" id="toc-preprocesamiento" class="nav-link" data-scroll-target="#preprocesamiento"><span class="header-section-number">3.1</span> Preprocesamiento</a></li>
  <li><a href="#particionamiento-de-datos" id="toc-particionamiento-de-datos" class="nav-link" data-scroll-target="#particionamiento-de-datos"><span class="header-section-number">3.2</span> Particionamiento de datos</a></li>
  <li><a href="#predicción-con-k-nn" id="toc-predicción-con-k-nn" class="nav-link" data-scroll-target="#predicción-con-k-nn"><span class="header-section-number">3.3</span> Predicción con k-NN</a></li>
  <li><a href="#análisis-y-validación" id="toc-análisis-y-validación" class="nav-link" data-scroll-target="#análisis-y-validación"><span class="header-section-number">3.4</span> Análisis y validación</a></li>
  </ul></li>
  <li><a href="#clasificación-con-lda" id="toc-clasificación-con-lda" class="nav-link" data-scroll-target="#clasificación-con-lda"><span class="header-section-number">4</span> Clasificación con LDA</a>
  <ul class="collapse">
  <li><a href="#preprocesamiento-1" id="toc-preprocesamiento-1" class="nav-link" data-scroll-target="#preprocesamiento-1"><span class="header-section-number">4.1</span> Preprocesamiento</a></li>
  <li><a href="#particionamiento-de-datos-1" id="toc-particionamiento-de-datos-1" class="nav-link" data-scroll-target="#particionamiento-de-datos-1"><span class="header-section-number">4.2</span> Particionamiento de datos</a></li>
  <li><a href="#predicción-con-lda" id="toc-predicción-con-lda" class="nav-link" data-scroll-target="#predicción-con-lda"><span class="header-section-number">4.3</span> Predicción con LDA</a></li>
  <li><a href="#analisis-y-validación" id="toc-analisis-y-validación" class="nav-link" data-scroll-target="#analisis-y-validación"><span class="header-section-number">4.4</span> Analisis y validación</a></li>
  </ul></li>
  <li><a href="#clasificación-con-qda" id="toc-clasificación-con-qda" class="nav-link" data-scroll-target="#clasificación-con-qda"><span class="header-section-number">5</span> Clasificación con QDA</a></li>
  <li><a href="#lda-y-qda-en-caret" id="toc-lda-y-qda-en-caret" class="nav-link" data-scroll-target="#lda-y-qda-en-caret"><span class="header-section-number">6</span> LDA y QDA en <code>caret</code></a></li>
  </ul>
<div class="quarto-code-links"><h2>Enlaces de código</h2><ul><li><a href="https://github.com/jgromero/datcom_classification_24"><i class="bi bi-github"></i>datcom-classification</a></li></ul></div></nav>
</div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Clasificación con el conjunto de datos Iris</h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Código</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Mostrar todo el código</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Ocultar todo el código</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">Ver el código fuente</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Autor/a</div>
    <div class="quarto-title-meta-contents">
             <p>Juan Gómez Romero </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Fecha de publicación</div>
    <div class="quarto-title-meta-contents">
      <p class="date">24 de octubre de 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<p>Clasificación con el conjunto de datos <a href="https://www.rdocumentation.org/packages/datasets/versions/3.6.2/topics/iris">iris</a>.</p>
<blockquote class="blockquote">
<p>El conjunto de datos <em>flor Iris</em> o conjunto de datos <em>iris</em> de Fisher es un conjunto de datos multivariante introducido por Ronald Fisher en su artículo de 1936 <em>The use of multiple measurements in taxonomic problems</em> (El uso de medidas múltiples en problemas taxonómicos) como un ejemplo de análisis discriminante lineal. El conjunto de datos contiene 50 muestras de cada una de tres especies de Iris (Iris setosa, Iris virginica e Iris versicolor). Se midieron cuatro rasgos de cada muestra: el largo y ancho del sépalo y pétalo, en centímetros. Basado en la combinación de estos cuatro rasgos, Fisher desarrolló un modelo discriminante lineal para distinguir entre una especie y otra (Fuente: <a href="https://es.wikipedia.org/wiki/Conjunto_de_datos_flor_iris">Wikipedia</a>).</p>
</blockquote>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="iris.png" class="img-fluid figure-img"></p>
<figcaption>Especies de la flor iris</figcaption>
</figure>
</div>
<section id="lectura-de-datos" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Lectura de datos</h1>
<p>Cargamos el conjunto de datos iris desde <a href="https://rdocumentation.org/packages/datasets/versions/3.6.2"><code>datasets</code></a>.</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>iris_data <span class="ot">&lt;-</span> iris</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(iris_data)</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":[""],"name":["_rn_"],"type":[""],"align":["left"]},{"label":["Sepal.Length"],"name":[1],"type":["dbl"],"align":["right"]},{"label":["Sepal.Width"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["Petal.Length"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["Petal.Width"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Species"],"name":[5],"type":["fct"],"align":["left"]}],"data":[{"1":"5.1","2":"3.5","3":"1.4","4":"0.2","5":"setosa","_rn_":"1"},{"1":"4.9","2":"3.0","3":"1.4","4":"0.2","5":"setosa","_rn_":"2"},{"1":"4.7","2":"3.2","3":"1.3","4":"0.2","5":"setosa","_rn_":"3"},{"1":"4.6","2":"3.1","3":"1.5","4":"0.2","5":"setosa","_rn_":"4"},{"1":"5.0","2":"3.6","3":"1.4","4":"0.2","5":"setosa","_rn_":"5"},{"1":"5.4","2":"3.9","3":"1.7","4":"0.4","5":"setosa","_rn_":"6"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
</section>
<section id="análisis-exploratorio" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Análisis exploratorio</h1>
<section id="características" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="características"><span class="header-section-number">2.1</span> Características</h2>
<p>Resumen estadístico básico de los datos: mínimos, máximos, medias, medianas, cuartiles, etc.</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(iris_data)</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>  Sepal.Length    Sepal.Width     Petal.Length    Petal.Width   
 Min.   :4.300   Min.   :2.000   Min.   :1.000   Min.   :0.100  
 1st Qu.:5.100   1st Qu.:2.800   1st Qu.:1.600   1st Qu.:0.300  
 Median :5.800   Median :3.000   Median :4.350   Median :1.300  
 Mean   :5.843   Mean   :3.057   Mean   :3.758   Mean   :1.199  
 3rd Qu.:6.400   3rd Qu.:3.300   3rd Qu.:5.100   3rd Qu.:1.800  
 Max.   :7.900   Max.   :4.400   Max.   :6.900   Max.   :2.500  
       Species  
 setosa    :50  
 versicolor:50  
 virginica :50  
                
                
                </code></pre>
</div>
</div>
<p>Los cuartiles se pueden ver gráficamente también, por ejemplo para <code>Sepal.Lenght</code>.</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(iris_data<span class="sc">$</span>Sepal.Length <span class="sc">~</span> iris_data<span class="sc">$</span>Species)</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="iris_files/figure-html/cuartiles-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="histograma" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="histograma"><span class="header-section-number">2.2</span> Histograma</h2>
<p>El histograma sirve para observar la distribución de los valores de una variable (cuáles se repiten y cómo).</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> iris_data) <span class="sc">+</span> </span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">x =</span> Sepal.Length), <span class="at">bins =</span> params<span class="sc">$</span>n_bins, </span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>                 <span class="at">color =</span> <span class="st">"white"</span>, <span class="at">fill =</span> <span class="st">"green"</span>)</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="iris_files/figure-html/histograma-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="densidad" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="densidad"><span class="header-section-number">2.3</span> Densidad</h2>
<p>La función de densidad es una versión suavizada del histograma, útil para datos continuos.</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> iris_data) <span class="sc">+</span> </span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="fu">aes</span>(<span class="at">x =</span> Sepal.Length),</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>               <span class="at">color =</span> <span class="st">"white"</span>, <span class="at">fill =</span> <span class="st">"lightgray"</span>)</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="iris_files/figure-html/densidad-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="dispersión" class="level2" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="dispersión"><span class="header-section-number">2.4</span> Dispersión</h2>
<p>La <a href="#fig-dispersion" class="quarto-xref">Figura&nbsp;1</a> muestra que las clases son claramente separables a partir de las variables de predicción.</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-dispersion" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-dispersion-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="iris_files/figure-html/fig-dispersion-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-dispersion-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figura&nbsp;1: Diagrama de dispersion.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Con <a href="https://www.rdocumentation.org/packages/GGally/versions/1.5.0/topics/ggpairs">ggpairs</a> podemos mostrar varios gráficos y métricas: correlación entre variables numéricas, densidades y diagramas de dispersión.</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggpairs</span>(iris_data, <span class="at">columns =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="fu">aes</span>(<span class="at">color =</span> Species, <span class="at">alpha =</span> <span class="fl">0.5</span>))</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in geom_point(): All aesthetics have length 1, but the data has 16 rows.
ℹ Did you mean to use `annotate()`?</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="iris_files/figure-html/pares-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="clasificación-con-k-nn" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Clasificación con k-NN</h1>
<section id="preprocesamiento" class="level2" data-number="3.1">
<h2 data-number="3.1" class="anchored" data-anchor-id="preprocesamiento"><span class="header-section-number">3.1</span> Preprocesamiento</h2>
<p>Normalizamos los datos para el cálculo de distancias con <a href="https://rdrr.io/rforge/caret/man/preProcess.html"><code>preProcess</code></a> .</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>sobj <span class="ot">&lt;-</span> <span class="fu">preProcess</span>(iris_data[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>], <span class="at">method=</span><span class="fu">c</span>(<span class="st">"scale"</span>))</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>iris_scaled <span class="ot">&lt;-</span> <span class="fu">predict</span>(sobj, iris_data[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]) <span class="sc">%&gt;%</span> <span class="fu">mutate</span>(<span class="at">Species =</span> iris_data<span class="sc">$</span>Species)</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="particionamiento-de-datos" class="level2" data-number="3.2">
<h2 data-number="3.2" class="anchored" data-anchor-id="particionamiento-de-datos"><span class="header-section-number">3.2</span> Particionamiento de datos</h2>
<p>Separamos las instancias que se usarán como <em>entrenamiento</em> de los individuos de validación usando <a href="https://rdrr.io/rforge/caret/man/createDataPartition.html"><code>createDataPartition</code></a> de <a href="http://topepo.github.io/caret/"><code>caret</code></a>.</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">0</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>trainIndex <span class="ot">&lt;-</span> <span class="fu">createDataPartition</span>(iris_scaled<span class="sc">$</span>Species, <span class="at">p =</span> .<span class="dv">50</span>, <span class="at">list =</span> <span class="cn">FALSE</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>train <span class="ot">&lt;-</span> iris_scaled[trainIndex, ] </span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>val   <span class="ot">&lt;-</span> iris_scaled[<span class="sc">-</span>trainIndex, ]</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="predicción-con-k-nn" class="level2" data-number="3.3">
<h2 data-number="3.3" class="anchored" data-anchor-id="predicción-con-k-nn"><span class="header-section-number">3.3</span> Predicción con k-NN</h2>
<p>Podemos aplicar k-NN para clasificar los datos de validación a partir de las instancias de entrenamiento.</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>knn.pred <span class="ot">&lt;-</span> <span class="fu">knn</span>(train[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>], val[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>], train<span class="sc">$</span>Species, <span class="at">k =</span> <span class="dv">5</span>)</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="análisis-y-validación" class="level2" data-number="3.4">
<h2 data-number="3.4" class="anchored" data-anchor-id="análisis-y-validación"><span class="header-section-number">3.4</span> Análisis y validación</h2>
<p>Calculadas las predicciones de los datos de validación, podemos comprobar los aciertos.</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>(t <span class="ot">&lt;-</span> <span class="fu">table</span>(knn.pred, val<span class="sc">$</span>Species))</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>            
knn.pred     setosa versicolor virginica
  setosa         25          0         0
  versicolor      0         22         1
  virginica       0          3        24</code></pre>
</div>
</div>
<p>Y calcular una tasa de acierto.</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>val_acc_rate <span class="ot">&lt;-</span> (t[<span class="dv">1</span>,<span class="dv">1</span>] <span class="sc">+</span> t[<span class="dv">2</span>,<span class="dv">2</span>] <span class="sc">+</span> t[<span class="dv">3</span>, <span class="dv">3</span>]) <span class="sc">/</span> <span class="fu">nrow</span>(val)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"% de acierto en validación: "</span>, val_acc_rate))</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "% de acierto en validación: 0.946666666666667"</code></pre>
</div>
</div>
<p>Podemos marcar sobre el gráfico de <code>Petal.Lenght</code> vs <code>Petal.Width</code> los valores conocidos y las predicciones.</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> val <span class="sc">%&gt;%</span> </span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">prediction =</span> knn.pred) <span class="sc">%&gt;%</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rename</span>(<span class="at">known =</span> Species) <span class="sc">%&gt;%</span></span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="fu">c</span>(<span class="st">"prediction"</span>, <span class="st">"known"</span>))</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> plot_data) <span class="sc">+</span> </span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> .data[[params<span class="sc">$</span>comparar_x]], <span class="at">y =</span> .data[[params<span class="sc">$</span>comparar_y]], <span class="at">shape =</span> name, <span class="at">color =</span> value))  <span class="sc">+</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> params<span class="sc">$</span>comparar_x, <span class="at">y =</span> params<span class="sc">$</span>comparar_y) <span class="sc">+</span></span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">name =</span> <span class="st">"Species"</span>, <span class="at">values=</span><span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>), <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"Iris Setosa"</span>, <span class="st">"Iris Versicolor"</span>, <span class="st">"Iris Virginica"</span>)) <span class="sc">+</span> </span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_shape_manual</span>(<span class="at">name =</span> <span class="st">"Type"</span>, <span class="at">values=</span><span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">3</span>))</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="iris_files/figure-html/validacion_plot-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="clasificación-con-lda" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Clasificación con LDA</h1>
<section id="preprocesamiento-1" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="preprocesamiento-1"><span class="header-section-number">4.1</span> Preprocesamiento</h2>
<p>El primer paso con LDA es comprobar las hipótesis de partida. Más en detalle:</p>
<p>– Que las observaciones siguen una distribución normal para cada predictor.</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> iris_data <span class="sc">%&gt;%</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span>Species)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> plot_data) <span class="sc">+</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="fu">aes</span>(<span class="at">x=</span>value, <span class="at">fill=</span>name)) <span class="sc">+</span> </span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_wrap</span>(<span class="sc">~</span> Species <span class="sc">+</span> name, <span class="at">scales =</span> <span class="st">"free"</span>) <span class="sc">+</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Density functions for each predictor and target value"</span>) <span class="sc">+</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_ipsum</span>(<span class="at">name =</span><span class="st">"Predictor"</span>) <span class="sc">+</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_ipsum_rc</span>()</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="iris_files/figure-html/lda_hipotesis_normal-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Cuantitativamente, podemos realizar el test de normalidad de Shapiro-Wilk con <a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/shapiro.test"><code>shapiro.test</code></a>.</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>my_shapiro <span class="ot">&lt;-</span> <span class="cf">function</span>(a) {</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">shapiro.test</span>(a)<span class="sc">$</span>statistic</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>iris_data <span class="sc">%&gt;%</span></span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(Species) <span class="sc">%&gt;%</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="fu">across</span>(Sepal.Length<span class="sc">:</span>Petal.Width, my_shapiro))</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div data-pagedtable="false">
  <script data-pagedtable-source="" type="application/json">
{"columns":[{"label":["Species"],"name":[1],"type":["fct"],"align":["left"]},{"label":["Sepal.Length"],"name":[2],"type":["dbl"],"align":["right"]},{"label":["Sepal.Width"],"name":[3],"type":["dbl"],"align":["right"]},{"label":["Petal.Length"],"name":[4],"type":["dbl"],"align":["right"]},{"label":["Petal.Width"],"name":[5],"type":["dbl"],"align":["right"]}],"data":[{"1":"setosa","2":"0.9776985","3":"0.9717195","4":"0.9549768","5":"0.7997645"},{"1":"versicolor","2":"0.9778357","3":"0.9741333","4":"0.9660044","5":"0.9476263"},{"1":"virginica","2":"0.9711794","3":"0.9673905","4":"0.9621864","5":"0.9597715"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
</div>
<p>– Que las matrices de varianza-covarianza son homogéneas.</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> iris_data <span class="sc">%&gt;%</span> </span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="sc">-</span>Species)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(plot_data) <span class="sc">+</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>       <span class="fu">geom_boxplot</span>(<span class="fu">aes</span>(<span class="at">x=</span>name, <span class="at">y=</span>value, <span class="at">fill=</span>Species)) <span class="sc">+</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>       <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Boxplots for each predictor and target value"</span>) <span class="sc">+</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>       <span class="fu">scale_fill_ipsum</span>(<span class="at">name =</span><span class="st">"Species"</span>) <span class="sc">+</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>       <span class="fu">theme_ipsum_rc</span>()</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="iris_files/figure-html/lda_hipotesis_homogen-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Cuantitativamente, podemos calcular las matrices de covarianza o aplicar el test de Bartlett con <a href="https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/bartlett.test"><code>bartlett.test</code></a>.</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>iris_data <span class="sc">%&gt;%</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(Species) <span class="sc">%&gt;%</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_map</span>(<span class="sc">~</span><span class="fu">var</span>(.x))</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
             Sepal.Length Sepal.Width Petal.Length Petal.Width
Sepal.Length   0.12424898 0.099216327  0.016355102 0.010330612
Sepal.Width    0.09921633 0.143689796  0.011697959 0.009297959
Petal.Length   0.01635510 0.011697959  0.030159184 0.006069388
Petal.Width    0.01033061 0.009297959  0.006069388 0.011106122

[[2]]
             Sepal.Length Sepal.Width Petal.Length Petal.Width
Sepal.Length   0.26643265  0.08518367   0.18289796  0.05577959
Sepal.Width    0.08518367  0.09846939   0.08265306  0.04120408
Petal.Length   0.18289796  0.08265306   0.22081633  0.07310204
Petal.Width    0.05577959  0.04120408   0.07310204  0.03910612

[[3]]
             Sepal.Length Sepal.Width Petal.Length Petal.Width
Sepal.Length   0.40434286  0.09376327   0.30328980  0.04909388
Sepal.Width    0.09376327  0.10400408   0.07137959  0.04762857
Petal.Length   0.30328980  0.07137959   0.30458776  0.04882449
Petal.Width    0.04909388  0.04762857   0.04882449  0.07543265</code></pre>
</div>
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bartlett.test</span>(Sepal.Length <span class="sc">~</span> Species, iris_data)</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
    Bartlett test of homogeneity of variances

data:  Sepal.Length by Species
Bartlett's K-squared = 16.006, df = 2, p-value = 0.0003345</code></pre>
</div>
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bartlett.test</span>(Sepal.Width  <span class="sc">~</span> Species, iris_data)</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
    Bartlett test of homogeneity of variances

data:  Sepal.Width by Species
Bartlett's K-squared = 2.0911, df = 2, p-value = 0.3515</code></pre>
</div>
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bartlett.test</span>(Petal.Length <span class="sc">~</span> Species, iris_data)</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
    Bartlett test of homogeneity of variances

data:  Petal.Length by Species
Bartlett's K-squared = 55.423, df = 2, p-value = 9.229e-13</code></pre>
</div>
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bartlett.test</span>(Petal.Width  <span class="sc">~</span> Species, iris_data)</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
    Bartlett test of homogeneity of variances

data:  Petal.Width by Species
Bartlett's K-squared = 39.213, df = 2, p-value = 3.055e-09</code></pre>
</div>
</div>
</section>
<section id="particionamiento-de-datos-1" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="particionamiento-de-datos-1"><span class="header-section-number">4.2</span> Particionamiento de datos</h2>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">0</span>)</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>trainIndex <span class="ot">&lt;-</span> <span class="fu">createDataPartition</span>(iris_scaled<span class="sc">$</span>Species, <span class="at">p =</span> .<span class="dv">50</span>, <span class="at">list =</span> <span class="cn">FALSE</span>)</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>train <span class="ot">&lt;-</span> iris_data[trainIndex, ] </span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>val   <span class="ot">&lt;-</span> iris_data[<span class="sc">-</span>trainIndex, ]</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="predicción-con-lda" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="predicción-con-lda"><span class="header-section-number">4.3</span> Predicción con LDA</h2>
<p>Podemos crear el modelo LDA para clasificar los datos de validación a partir de las instancias de entrenamiento.</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>model.lda <span class="ot">&lt;-</span> <span class="fu">lda</span>(Species <span class="sc">~</span> Sepal.Length <span class="sc">+</span> Sepal.Width <span class="sc">+</span> Petal.Length <span class="sc">+</span> Petal.Width, <span class="at">data =</span> train)</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>model.lda</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Call:
lda(Species ~ Sepal.Length + Sepal.Width + Petal.Length + Petal.Width, 
    data = train)

Prior probabilities of groups:
    setosa versicolor  virginica 
 0.3333333  0.3333333  0.3333333 

Group means:
           Sepal.Length Sepal.Width Petal.Length Petal.Width
setosa            4.888       3.400        1.424       0.212
versicolor        5.864       2.756        4.188       1.284
virginica         6.548       2.996        5.484       1.996

Coefficients of linear discriminants:
                    LD1        LD2
Sepal.Length  0.5966349 -0.3505672
Sepal.Width   1.7195798 -2.1521860
Petal.Length -1.7643180  1.3264664
Petal.Width  -3.5360050 -3.3672017

Proportion of trace:
   LD1    LD2 
0.9878 0.0122 </code></pre>
</div>
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>lda.pred.train <span class="ot">&lt;-</span> <span class="fu">predict</span>(model.lda, train[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>])</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>lda.pred.val <span class="ot">&lt;-</span> <span class="fu">predict</span>(model.lda, val[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>])</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Observar la transformación de los datos usados para crear el modelo.</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> lda.pred.train<span class="sc">$</span>x <span class="sc">%&gt;%</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">Species =</span> train[[<span class="dv">5</span>]])</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> plot_data) <span class="sc">+</span> </span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> LD1, <span class="at">y =</span> LD2, <span class="at">color =</span> Species))  <span class="sc">+</span></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">name =</span> <span class="st">"Species"</span>, <span class="at">values=</span><span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>), <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"Iris Setosa"</span>, <span class="st">"Iris Versicolor"</span>, <span class="st">"Iris Virginica"</span>)) <span class="sc">%&gt;%</span></span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Data transformed after LDA"</span>) <span class="sc">+</span></span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_ipsum_rc</span>()</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="iris_files/figure-html/lda_transformacion-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="analisis-y-validación" class="level2" data-number="4.4">
<h2 data-number="4.4" class="anchored" data-anchor-id="analisis-y-validación"><span class="header-section-number">4.4</span> Analisis y validación</h2>
<p>Finalmente, podemos calcular las métricas en validación.</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>(t <span class="ot">&lt;-</span> <span class="fu">table</span>(lda.pred.val<span class="sc">$</span>class, val<span class="sc">$</span>Species))</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>            
             setosa versicolor virginica
  setosa         25          0         0
  versicolor      0         23         0
  virginica       0          2        25</code></pre>
</div>
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>val_acc_rate <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="fu">diag</span>(t)) <span class="sc">/</span> <span class="fu">nrow</span>(val)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"% de acierto en validación: "</span>, val_acc_rate))</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "% de acierto en validación: 0.973333333333333"</code></pre>
</div>
</div>
<p>Y visualizar las predicciones, tanto con los datos de entrenamiento como los de validación.</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> lda.pred.train<span class="sc">$</span>x <span class="sc">%&gt;%</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">known =</span> train[[<span class="dv">5</span>]]) <span class="sc">%&gt;%</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">prediction =</span> lda.pred.train<span class="sc">$</span>class) <span class="sc">%&gt;%</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="fu">c</span>(<span class="st">"prediction"</span>, <span class="st">"known"</span>))</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> plot_data) <span class="sc">+</span> </span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> LD1, <span class="at">y =</span> LD2, <span class="at">shape =</span> name, <span class="at">color =</span> value))  <span class="sc">+</span></span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">name =</span> <span class="st">"Species"</span>, <span class="at">values=</span><span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>), <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"Iris Setosa"</span>, <span class="st">"Iris Versicolor"</span>, <span class="st">"Iris Virginica"</span>)) <span class="sc">+</span> </span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_shape_manual</span>(<span class="at">name =</span> <span class="st">"Type"</span>, <span class="at">values=</span><span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">3</span>)) <span class="sc">+</span></span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Train data + predictions transformed after LDA"</span>) <span class="sc">+</span></span>
<span id="cb39-12"><a href="#cb39-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_ipsum_rc</span>()</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="iris_files/figure-html/lda_prediccion_train-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>plot_data <span class="ot">&lt;-</span> lda.pred.val<span class="sc">$</span>x <span class="sc">%&gt;%</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">known =</span> val[[<span class="dv">5</span>]]) <span class="sc">%&gt;%</span></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">prediction =</span> lda.pred.val<span class="sc">$</span>class) <span class="sc">%&gt;%</span></span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="fu">c</span>(<span class="st">"prediction"</span>, <span class="st">"known"</span>))</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> plot_data) <span class="sc">+</span> </span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="fu">aes</span>(<span class="at">x =</span> LD1, <span class="at">y =</span> LD2, <span class="at">shape =</span> name, <span class="at">color =</span> value))  <span class="sc">+</span></span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_manual</span>(<span class="at">name =</span> <span class="st">"Species"</span>, <span class="at">values=</span><span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>), <span class="at">labels=</span><span class="fu">c</span>(<span class="st">"Iris Setosa"</span>, <span class="st">"Iris Versicolor"</span>, <span class="st">"Iris Virginica"</span>)) <span class="sc">+</span> </span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_shape_manual</span>(<span class="at">name =</span> <span class="st">"Type"</span>, <span class="at">values=</span><span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">3</span>)) <span class="sc">+</span></span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Validation data + predictions transformed after LDA"</span>) <span class="sc">+</span></span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_ipsum_rc</span>()</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="iris_files/figure-html/lda_prediccion_val-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="clasificación-con-qda" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Clasificación con QDA</h1>
<p>Por último, creamos un clasificador con QDA usando el método <a href="https://www.rdocumentation.org/packages/MASS/versions/7.3-58.3/topics/qda"><code>qda</code></a> de <a href="https://www.rdocumentation.org/packages/MASS/versions/7.3-58.3"><code>MASS</code></a>.</p>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>model.qda <span class="ot">&lt;-</span> <span class="fu">qda</span>(Species <span class="sc">~</span> Sepal.Length <span class="sc">+</span> Sepal.Width <span class="sc">+</span> Petal.Length <span class="sc">+</span> Petal.Width, <span class="at">data =</span> train)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>qda.pred.train <span class="ot">&lt;-</span> <span class="fu">predict</span>(model.qda, train[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>])</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>qda.pred.val <span class="ot">&lt;-</span> <span class="fu">predict</span>(model.qda, val[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>])</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>(t <span class="ot">&lt;-</span> <span class="fu">table</span>(qda.pred.val<span class="sc">$</span>class, val<span class="sc">$</span>Species))</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>            
             setosa versicolor virginica
  setosa         25          0         0
  versicolor      0         22         0
  virginica       0          3        25</code></pre>
</div>
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>val_acc_rate <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="fu">diag</span>(t)) <span class="sc">/</span> <span class="fu">nrow</span>(val)</span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"% de acierto en validación: "</span>, val_acc_rate))</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "% de acierto en validación: 0.96"</code></pre>
</div>
</div>
</section>
<section id="lda-y-qda-en-caret" class="level1" data-number="6">
<h1 data-number="6"><span class="header-section-number">6</span> LDA y QDA en <code>caret</code></h1>
<div class="cell">
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressWarnings</span>({</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>  lda.model <span class="ot">&lt;-</span> <span class="fu">train</span>(train[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>],</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>                    train<span class="sc">$</span>Species,</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">method=</span><span class="st">"lda"</span>, </span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>                    <span class="at">preProcess =</span> <span class="fu">c</span>(<span class="st">"center"</span>, <span class="st">"scale"</span>),</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>                    <span class="at">tuneLength =</span> <span class="dv">10</span>,</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>                    <span class="at">trControl =</span> <span class="fu">trainControl</span>(<span class="at">method =</span> <span class="st">"cv"</span>))</span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>lda.model</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Linear Discriminant Analysis 

75 samples
 4 predictor
 3 classes: 'setosa', 'versicolor', 'virginica' 

Pre-processing: centered (4), scaled (4) 
Resampling: Cross-Validated (10 fold) 
Summary of sample sizes: 67, 68, 67, 67, 69, 68, ... 
Resampling results:

  Accuracy  Kappa    
  0.975     0.9619048</code></pre>
</div>
<details class="code-fold">
<summary>Código</summary>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressWarnings</span>({</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a>  qda.model <span class="ot">&lt;-</span> <span class="fu">train</span>(train[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>],</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>                    train<span class="sc">$</span>Species,</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>                    <span class="at">method=</span><span class="st">"qda"</span>, </span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>                    <span class="at">preProcess =</span> <span class="fu">c</span>(<span class="st">"center"</span>, <span class="st">"scale"</span>),</span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>                    <span class="at">tuneLength =</span> <span class="dv">10</span>,</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a>                    <span class="at">trControl =</span> <span class="fu">trainControl</span>(<span class="at">method =</span> <span class="st">"cv"</span>))</span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a>qda.model</span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Quadratic Discriminant Analysis 

75 samples
 4 predictor
 3 classes: 'setosa', 'versicolor', 'virginica' 

Pre-processing: centered (4), scaled (4) 
Resampling: Cross-Validated (10 fold) 
Summary of sample sizes: 67, 69, 68, 67, 67, 69, ... 
Resampling results:

  Accuracy   Kappa    
  0.9315476  0.8952549</code></pre>
</div>
</div>
<!-- -->

</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copiado");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copiado");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Ejecutar el código</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb49" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Clasificación con el conjunto de datos Iris"</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> "Juan Gómez Romero"</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> "10/24/2024"</span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a><span class="an">lang:</span><span class="co"> es</span></span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a><span class="co">  html:</span></span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools: true</span></span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a><span class="co">    code-fold: true</span></span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a><span class="co">    toc: true</span></span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a><span class="co">    number-sections: true</span></span>
<span id="cb49-12"><a href="#cb49-12" aria-hidden="true" tabindex="-1"></a><span class="co">    css: styles.css</span></span>
<span id="cb49-13"><a href="#cb49-13" aria-hidden="true" tabindex="-1"></a><span class="co">    df-print: paged</span></span>
<span id="cb49-14"><a href="#cb49-14" aria-hidden="true" tabindex="-1"></a><span class="co">    code-links:</span></span>
<span id="cb49-15"><a href="#cb49-15" aria-hidden="true" tabindex="-1"></a><span class="co">      - text: datcom-classification</span></span>
<span id="cb49-16"><a href="#cb49-16" aria-hidden="true" tabindex="-1"></a><span class="co">        icon: github</span></span>
<span id="cb49-17"><a href="#cb49-17" aria-hidden="true" tabindex="-1"></a><span class="co">        href: https://github.com/jgromero/datcom_classification_24 </span></span>
<span id="cb49-18"><a href="#cb49-18" aria-hidden="true" tabindex="-1"></a><span class="an">params:</span></span>
<span id="cb49-19"><a href="#cb49-19" aria-hidden="true" tabindex="-1"></a><span class="co">  n_bins:</span></span>
<span id="cb49-20"><a href="#cb49-20" aria-hidden="true" tabindex="-1"></a><span class="co">    value: 10</span></span>
<span id="cb49-21"><a href="#cb49-21" aria-hidden="true" tabindex="-1"></a><span class="co">  comparar_x:</span></span>
<span id="cb49-22"><a href="#cb49-22" aria-hidden="true" tabindex="-1"></a><span class="co">    value: "Petal.Length"</span></span>
<span id="cb49-23"><a href="#cb49-23" aria-hidden="true" tabindex="-1"></a><span class="co">  comparar_y:</span></span>
<span id="cb49-24"><a href="#cb49-24" aria-hidden="true" tabindex="-1"></a><span class="co">    value: "Petal.Width"</span></span>
<span id="cb49-25"><a href="#cb49-25" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb49-26"><a href="#cb49-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-27"><a href="#cb49-27" aria-hidden="true" tabindex="-1"></a><span class="in">```{r setup, include=FALSE}</span></span>
<span id="cb49-28"><a href="#cb49-28" aria-hidden="true" tabindex="-1"></a><span class="in">knitr::opts_chunk$set(message = FALSE)</span></span>
<span id="cb49-29"><a href="#cb49-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-30"><a href="#cb49-30" aria-hidden="true" tabindex="-1"></a><span class="in">library(knitr)</span></span>
<span id="cb49-31"><a href="#cb49-31" aria-hidden="true" tabindex="-1"></a><span class="in">library(tidyverse)</span></span>
<span id="cb49-32"><a href="#cb49-32" aria-hidden="true" tabindex="-1"></a><span class="in">library(datasets)</span></span>
<span id="cb49-33"><a href="#cb49-33" aria-hidden="true" tabindex="-1"></a><span class="in">library(caret)</span></span>
<span id="cb49-34"><a href="#cb49-34" aria-hidden="true" tabindex="-1"></a><span class="in">library(class)</span></span>
<span id="cb49-35"><a href="#cb49-35" aria-hidden="true" tabindex="-1"></a><span class="in">library(hrbrthemes)</span></span>
<span id="cb49-36"><a href="#cb49-36" aria-hidden="true" tabindex="-1"></a><span class="in">library(MASS)</span></span>
<span id="cb49-37"><a href="#cb49-37" aria-hidden="true" tabindex="-1"></a><span class="in">library(GGally)</span></span>
<span id="cb49-38"><a href="#cb49-38" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-39"><a href="#cb49-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-40"><a href="#cb49-40" aria-hidden="true" tabindex="-1"></a>Clasificación con el conjunto de datos <span class="co">[</span><span class="ot">iris</span><span class="co">](https://www.rdocumentation.org/packages/datasets/versions/3.6.2/topics/iris)</span>.</span>
<span id="cb49-41"><a href="#cb49-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-42"><a href="#cb49-42" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; El conjunto de datos *flor Iris* o conjunto de datos *iris* de Fisher es un conjunto de datos multivariante introducido por Ronald Fisher en su artículo de 1936 *The use of multiple measurements in taxonomic problems* (El uso de medidas múltiples en problemas taxonómicos) como un ejemplo de análisis discriminante lineal. El conjunto de datos contiene 50 muestras de cada una de tres especies de Iris (Iris setosa, Iris virginica e Iris versicolor). Se midieron cuatro rasgos de cada muestra: el largo y ancho del sépalo y pétalo, en centímetros. Basado en la combinación de estos cuatro rasgos, Fisher desarrolló un modelo discriminante lineal para distinguir entre una especie y otra (Fuente: </span><span class="co">[</span><span class="ot">Wikipedia</span><span class="co">](https://es.wikipedia.org/wiki/Conjunto_de_datos_flor_iris)</span><span class="at">).</span></span>
<span id="cb49-43"><a href="#cb49-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-44"><a href="#cb49-44" aria-hidden="true" tabindex="-1"></a><span class="al">![Especies de la flor iris](iris.png)</span></span>
<span id="cb49-45"><a href="#cb49-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-46"><a href="#cb49-46" aria-hidden="true" tabindex="-1"></a><span class="fu"># Lectura de datos</span></span>
<span id="cb49-47"><a href="#cb49-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-48"><a href="#cb49-48" aria-hidden="true" tabindex="-1"></a>Cargamos el conjunto de datos iris desde <span class="co">[</span><span class="ot">`datasets`</span><span class="co">](https://rdocumentation.org/packages/datasets/versions/3.6.2)</span>.</span>
<span id="cb49-49"><a href="#cb49-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-50"><a href="#cb49-50" aria-hidden="true" tabindex="-1"></a><span class="in">```{r lectura}</span></span>
<span id="cb49-51"><a href="#cb49-51" aria-hidden="true" tabindex="-1"></a><span class="in">iris_data &lt;- iris</span></span>
<span id="cb49-52"><a href="#cb49-52" aria-hidden="true" tabindex="-1"></a><span class="in">head(iris_data)</span></span>
<span id="cb49-53"><a href="#cb49-53" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-54"><a href="#cb49-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-55"><a href="#cb49-55" aria-hidden="true" tabindex="-1"></a><span class="fu"># Análisis exploratorio</span></span>
<span id="cb49-56"><a href="#cb49-56" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-57"><a href="#cb49-57" aria-hidden="true" tabindex="-1"></a><span class="fu">## Características</span></span>
<span id="cb49-58"><a href="#cb49-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-59"><a href="#cb49-59" aria-hidden="true" tabindex="-1"></a>Resumen estadístico básico de los datos: mínimos, máximos, medias, medianas, cuartiles, etc.</span>
<span id="cb49-60"><a href="#cb49-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-61"><a href="#cb49-61" aria-hidden="true" tabindex="-1"></a><span class="in">```{r resumen}</span></span>
<span id="cb49-62"><a href="#cb49-62" aria-hidden="true" tabindex="-1"></a><span class="in">summary(iris_data)</span></span>
<span id="cb49-63"><a href="#cb49-63" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-64"><a href="#cb49-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-65"><a href="#cb49-65" aria-hidden="true" tabindex="-1"></a>Los cuartiles se pueden ver gráficamente también, por ejemplo para <span class="in">`Sepal.Lenght`</span>.</span>
<span id="cb49-66"><a href="#cb49-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-67"><a href="#cb49-67" aria-hidden="true" tabindex="-1"></a><span class="in">```{r cuartiles}</span></span>
<span id="cb49-68"><a href="#cb49-68" aria-hidden="true" tabindex="-1"></a><span class="in">boxplot(iris_data$Sepal.Length ~ iris_data$Species)</span></span>
<span id="cb49-69"><a href="#cb49-69" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-70"><a href="#cb49-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-71"><a href="#cb49-71" aria-hidden="true" tabindex="-1"></a><span class="fu">## Histograma</span></span>
<span id="cb49-72"><a href="#cb49-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-73"><a href="#cb49-73" aria-hidden="true" tabindex="-1"></a>El histograma sirve para observar la distribución de los valores de una variable (cuáles se repiten y cómo).</span>
<span id="cb49-74"><a href="#cb49-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-75"><a href="#cb49-75" aria-hidden="true" tabindex="-1"></a><span class="in">```{r histograma}</span></span>
<span id="cb49-76"><a href="#cb49-76" aria-hidden="true" tabindex="-1"></a><span class="in">ggplot(data = iris_data) + </span></span>
<span id="cb49-77"><a href="#cb49-77" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_histogram(aes(x = Sepal.Length), bins = params$n_bins, </span></span>
<span id="cb49-78"><a href="#cb49-78" aria-hidden="true" tabindex="-1"></a><span class="in">                 color = "white", fill = "green")</span></span>
<span id="cb49-79"><a href="#cb49-79" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-80"><a href="#cb49-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-81"><a href="#cb49-81" aria-hidden="true" tabindex="-1"></a><span class="fu">## Densidad</span></span>
<span id="cb49-82"><a href="#cb49-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-83"><a href="#cb49-83" aria-hidden="true" tabindex="-1"></a>La función de densidad es una versión suavizada del histograma, útil para datos continuos.</span>
<span id="cb49-84"><a href="#cb49-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-85"><a href="#cb49-85" aria-hidden="true" tabindex="-1"></a><span class="in">```{r densidad}</span></span>
<span id="cb49-86"><a href="#cb49-86" aria-hidden="true" tabindex="-1"></a><span class="in">ggplot(data = iris_data) + </span></span>
<span id="cb49-87"><a href="#cb49-87" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_density(aes(x = Sepal.Length),</span></span>
<span id="cb49-88"><a href="#cb49-88" aria-hidden="true" tabindex="-1"></a><span class="in">               color = "white", fill = "lightgray")</span></span>
<span id="cb49-89"><a href="#cb49-89" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-90"><a href="#cb49-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-91"><a href="#cb49-91" aria-hidden="true" tabindex="-1"></a><span class="fu">## Dispersión</span></span>
<span id="cb49-92"><a href="#cb49-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-93"><a href="#cb49-93" aria-hidden="true" tabindex="-1"></a>La @fig-dispersion muestra que las clases son claramente separables a partir de las variables de predicción.</span>
<span id="cb49-94"><a href="#cb49-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-95"><a href="#cb49-95" aria-hidden="true" tabindex="-1"></a><span class="in">```{r dispersion, echo=FALSE}</span></span>
<span id="cb49-96"><a href="#cb49-96" aria-hidden="true" tabindex="-1"></a><span class="in">#| label: fig-dispersion</span></span>
<span id="cb49-97"><a href="#cb49-97" aria-hidden="true" tabindex="-1"></a><span class="in">#| fig-cap: "Diagrama de dispersion."</span></span>
<span id="cb49-98"><a href="#cb49-98" aria-hidden="true" tabindex="-1"></a><span class="in">#| warning: false</span></span>
<span id="cb49-99"><a href="#cb49-99" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-100"><a href="#cb49-100" aria-hidden="true" tabindex="-1"></a><span class="in">ggplot(data = iris_data) + </span></span>
<span id="cb49-101"><a href="#cb49-101" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_point(aes(x = .data[[params$comparar_x]], y = .data[[params$comparar_y]], color = Species, shape = Species))  +</span></span>
<span id="cb49-102"><a href="#cb49-102" aria-hidden="true" tabindex="-1"></a><span class="in">  labs(x = params$comparar_x, y = params$comparar_y) +  </span></span>
<span id="cb49-103"><a href="#cb49-103" aria-hidden="true" tabindex="-1"></a><span class="in">  scale_color_discrete(name ="Especie", labels=c("Iris Setosa", "Iris Versicolor", "Iris Virginica")) +</span></span>
<span id="cb49-104"><a href="#cb49-104" aria-hidden="true" tabindex="-1"></a><span class="in">  scale_shape_discrete(name ="Especie", labels=c("Iris Setosa", "Iris Versicolor", "Iris Virginica"))</span></span>
<span id="cb49-105"><a href="#cb49-105" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-106"><a href="#cb49-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-107"><a href="#cb49-107" aria-hidden="true" tabindex="-1"></a>Con <span class="co">[</span><span class="ot">ggpairs</span><span class="co">](https://www.rdocumentation.org/packages/GGally/versions/1.5.0/topics/ggpairs)</span> podemos mostrar varios gráficos y métricas: correlación entre variables numéricas, densidades y diagramas de dispersión.</span>
<span id="cb49-108"><a href="#cb49-108" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-109"><a href="#cb49-109" aria-hidden="true" tabindex="-1"></a><span class="in">```{r pares}</span></span>
<span id="cb49-110"><a href="#cb49-110" aria-hidden="true" tabindex="-1"></a><span class="in">ggpairs(iris_data, columns = 1:4, aes(color = Species, alpha = 0.5))</span></span>
<span id="cb49-111"><a href="#cb49-111" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-112"><a href="#cb49-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-113"><a href="#cb49-113" aria-hidden="true" tabindex="-1"></a><span class="fu"># Clasificación con k-NN</span></span>
<span id="cb49-114"><a href="#cb49-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-115"><a href="#cb49-115" aria-hidden="true" tabindex="-1"></a><span class="fu">## Preprocesamiento</span></span>
<span id="cb49-116"><a href="#cb49-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-117"><a href="#cb49-117" aria-hidden="true" tabindex="-1"></a>Normalizamos los datos para el cálculo de distancias con <span class="co">[</span><span class="ot">`preProcess`</span><span class="co">](https://rdrr.io/rforge/caret/man/preProcess.html)</span> .</span>
<span id="cb49-118"><a href="#cb49-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-119"><a href="#cb49-119" aria-hidden="true" tabindex="-1"></a><span class="in">```{r normalizar}</span></span>
<span id="cb49-120"><a href="#cb49-120" aria-hidden="true" tabindex="-1"></a><span class="in">sobj &lt;- preProcess(iris_data[1:4], method=c("scale"))</span></span>
<span id="cb49-121"><a href="#cb49-121" aria-hidden="true" tabindex="-1"></a><span class="in">iris_scaled &lt;- predict(sobj, iris_data[1:4]) %&gt;% mutate(Species = iris_data$Species)</span></span>
<span id="cb49-122"><a href="#cb49-122" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-123"><a href="#cb49-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-124"><a href="#cb49-124" aria-hidden="true" tabindex="-1"></a><span class="fu">## Particionamiento de datos</span></span>
<span id="cb49-125"><a href="#cb49-125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-126"><a href="#cb49-126" aria-hidden="true" tabindex="-1"></a>Separamos las instancias que se usarán como *entrenamiento* de los individuos de validación usando <span class="co">[</span><span class="ot">`createDataPartition`</span><span class="co">](https://rdrr.io/rforge/caret/man/createDataPartition.html)</span> de <span class="co">[</span><span class="ot">`caret`</span><span class="co">](http://topepo.github.io/caret/)</span>.</span>
<span id="cb49-127"><a href="#cb49-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-128"><a href="#cb49-128" aria-hidden="true" tabindex="-1"></a><span class="in">```{r particionamiento_knn}</span></span>
<span id="cb49-129"><a href="#cb49-129" aria-hidden="true" tabindex="-1"></a><span class="in">set.seed(0)</span></span>
<span id="cb49-130"><a href="#cb49-130" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-131"><a href="#cb49-131" aria-hidden="true" tabindex="-1"></a><span class="in">trainIndex &lt;- createDataPartition(iris_scaled$Species, p = .50, list = FALSE)</span></span>
<span id="cb49-132"><a href="#cb49-132" aria-hidden="true" tabindex="-1"></a><span class="in">train &lt;- iris_scaled[trainIndex, ] </span></span>
<span id="cb49-133"><a href="#cb49-133" aria-hidden="true" tabindex="-1"></a><span class="in">val   &lt;- iris_scaled[-trainIndex, ]</span></span>
<span id="cb49-134"><a href="#cb49-134" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-135"><a href="#cb49-135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-136"><a href="#cb49-136" aria-hidden="true" tabindex="-1"></a><span class="fu">## Predicción con k-NN</span></span>
<span id="cb49-137"><a href="#cb49-137" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-138"><a href="#cb49-138" aria-hidden="true" tabindex="-1"></a>Podemos aplicar k-NN para clasificar los datos de validación a partir de las instancias de entrenamiento.</span>
<span id="cb49-139"><a href="#cb49-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-140"><a href="#cb49-140" aria-hidden="true" tabindex="-1"></a><span class="in">```{r knn}</span></span>
<span id="cb49-141"><a href="#cb49-141" aria-hidden="true" tabindex="-1"></a><span class="in">knn.pred &lt;- knn(train[1:4], val[1:4], train$Species, k = 5)</span></span>
<span id="cb49-142"><a href="#cb49-142" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-143"><a href="#cb49-143" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-144"><a href="#cb49-144" aria-hidden="true" tabindex="-1"></a><span class="fu">## Análisis y validación</span></span>
<span id="cb49-145"><a href="#cb49-145" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-146"><a href="#cb49-146" aria-hidden="true" tabindex="-1"></a>Calculadas las predicciones de los datos de validación, podemos comprobar los aciertos.</span>
<span id="cb49-147"><a href="#cb49-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-148"><a href="#cb49-148" aria-hidden="true" tabindex="-1"></a><span class="in">```{r validacion}</span></span>
<span id="cb49-149"><a href="#cb49-149" aria-hidden="true" tabindex="-1"></a><span class="in">(t &lt;- table(knn.pred, val$Species))</span></span>
<span id="cb49-150"><a href="#cb49-150" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-151"><a href="#cb49-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-152"><a href="#cb49-152" aria-hidden="true" tabindex="-1"></a>Y calcular una tasa de acierto.</span>
<span id="cb49-153"><a href="#cb49-153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-154"><a href="#cb49-154" aria-hidden="true" tabindex="-1"></a><span class="in">```{r acierto}</span></span>
<span id="cb49-155"><a href="#cb49-155" aria-hidden="true" tabindex="-1"></a><span class="in">val_acc_rate &lt;- (t[1,1] + t[2,2] + t[3, 3]) / nrow(val)</span></span>
<span id="cb49-156"><a href="#cb49-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-157"><a href="#cb49-157" aria-hidden="true" tabindex="-1"></a><span class="in">print(paste0("% de acierto en validación: ", val_acc_rate))</span></span>
<span id="cb49-158"><a href="#cb49-158" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-159"><a href="#cb49-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-160"><a href="#cb49-160" aria-hidden="true" tabindex="-1"></a>Podemos marcar sobre el gráfico de <span class="in">`Petal.Lenght`</span> vs <span class="in">`Petal.Width`</span> los valores conocidos y las predicciones.</span>
<span id="cb49-161"><a href="#cb49-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-162"><a href="#cb49-162" aria-hidden="true" tabindex="-1"></a><span class="in">```{r validacion_plot}</span></span>
<span id="cb49-163"><a href="#cb49-163" aria-hidden="true" tabindex="-1"></a><span class="in">plot_data &lt;- val %&gt;% </span></span>
<span id="cb49-164"><a href="#cb49-164" aria-hidden="true" tabindex="-1"></a><span class="in">  mutate(prediction = knn.pred) %&gt;%</span></span>
<span id="cb49-165"><a href="#cb49-165" aria-hidden="true" tabindex="-1"></a><span class="in">  rename(known = Species) %&gt;%</span></span>
<span id="cb49-166"><a href="#cb49-166" aria-hidden="true" tabindex="-1"></a><span class="in">  pivot_longer(c("prediction", "known"))</span></span>
<span id="cb49-167"><a href="#cb49-167" aria-hidden="true" tabindex="-1"></a><span class="in">  </span></span>
<span id="cb49-168"><a href="#cb49-168" aria-hidden="true" tabindex="-1"></a><span class="in">ggplot(data = plot_data) + </span></span>
<span id="cb49-169"><a href="#cb49-169" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_point(aes(x = .data[[params$comparar_x]], y = .data[[params$comparar_y]], shape = name, color = value))  +</span></span>
<span id="cb49-170"><a href="#cb49-170" aria-hidden="true" tabindex="-1"></a><span class="in">  labs(x = params$comparar_x, y = params$comparar_y) +</span></span>
<span id="cb49-171"><a href="#cb49-171" aria-hidden="true" tabindex="-1"></a><span class="in">  scale_colour_manual(name = "Species", values=c("red", "green", "blue"), labels=c("Iris Setosa", "Iris Versicolor", "Iris Virginica")) + </span></span>
<span id="cb49-172"><a href="#cb49-172" aria-hidden="true" tabindex="-1"></a><span class="in">  scale_shape_manual(name = "Type", values=c(5, 3))</span></span>
<span id="cb49-173"><a href="#cb49-173" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-174"><a href="#cb49-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-175"><a href="#cb49-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-176"><a href="#cb49-176" aria-hidden="true" tabindex="-1"></a><span class="fu"># Clasificación con LDA</span></span>
<span id="cb49-177"><a href="#cb49-177" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-178"><a href="#cb49-178" aria-hidden="true" tabindex="-1"></a><span class="fu">## Preprocesamiento</span></span>
<span id="cb49-179"><a href="#cb49-179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-180"><a href="#cb49-180" aria-hidden="true" tabindex="-1"></a>El primer paso con LDA es comprobar las hipótesis de partida. Más en detalle:</span>
<span id="cb49-181"><a href="#cb49-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-182"><a href="#cb49-182" aria-hidden="true" tabindex="-1"></a>-- Que las observaciones siguen una distribución normal para cada predictor.</span>
<span id="cb49-183"><a href="#cb49-183" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-184"><a href="#cb49-184" aria-hidden="true" tabindex="-1"></a><span class="in">```{r lda_hipotesis_normal}</span></span>
<span id="cb49-185"><a href="#cb49-185" aria-hidden="true" tabindex="-1"></a><span class="in">plot_data &lt;- iris_data %&gt;%</span></span>
<span id="cb49-186"><a href="#cb49-186" aria-hidden="true" tabindex="-1"></a><span class="in">  pivot_longer(-Species)</span></span>
<span id="cb49-187"><a href="#cb49-187" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-188"><a href="#cb49-188" aria-hidden="true" tabindex="-1"></a><span class="in">ggplot(data = plot_data) +</span></span>
<span id="cb49-189"><a href="#cb49-189" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_density(aes(x=value, fill=name)) + </span></span>
<span id="cb49-190"><a href="#cb49-190" aria-hidden="true" tabindex="-1"></a><span class="in">  facet_wrap(~ Species + name, scales = "free") +</span></span>
<span id="cb49-191"><a href="#cb49-191" aria-hidden="true" tabindex="-1"></a><span class="in">  labs(title = "Density functions for each predictor and target value") +</span></span>
<span id="cb49-192"><a href="#cb49-192" aria-hidden="true" tabindex="-1"></a><span class="in">  scale_fill_ipsum(name ="Predictor") +</span></span>
<span id="cb49-193"><a href="#cb49-193" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_ipsum_rc()</span></span>
<span id="cb49-194"><a href="#cb49-194" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-195"><a href="#cb49-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-196"><a href="#cb49-196" aria-hidden="true" tabindex="-1"></a>Cuantitativamente, podemos realizar el test de normalidad de Shapiro-Wilk con <span class="co">[</span><span class="ot">`shapiro.test`</span><span class="co">](https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/shapiro.test)</span>.</span>
<span id="cb49-197"><a href="#cb49-197" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-198"><a href="#cb49-198" aria-hidden="true" tabindex="-1"></a><span class="in">```{r lda_shapiro}</span></span>
<span id="cb49-199"><a href="#cb49-199" aria-hidden="true" tabindex="-1"></a><span class="in">my_shapiro &lt;- function(a) {</span></span>
<span id="cb49-200"><a href="#cb49-200" aria-hidden="true" tabindex="-1"></a><span class="in">  shapiro.test(a)$statistic</span></span>
<span id="cb49-201"><a href="#cb49-201" aria-hidden="true" tabindex="-1"></a><span class="in">}</span></span>
<span id="cb49-202"><a href="#cb49-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-203"><a href="#cb49-203" aria-hidden="true" tabindex="-1"></a><span class="in">iris_data %&gt;%</span></span>
<span id="cb49-204"><a href="#cb49-204" aria-hidden="true" tabindex="-1"></a><span class="in">  group_by(Species) %&gt;%</span></span>
<span id="cb49-205"><a href="#cb49-205" aria-hidden="true" tabindex="-1"></a><span class="in">  summarise(across(Sepal.Length:Petal.Width, my_shapiro))</span></span>
<span id="cb49-206"><a href="#cb49-206" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-207"><a href="#cb49-207" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-208"><a href="#cb49-208" aria-hidden="true" tabindex="-1"></a>-- Que las matrices de varianza-covarianza son homogéneas.</span>
<span id="cb49-209"><a href="#cb49-209" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-210"><a href="#cb49-210" aria-hidden="true" tabindex="-1"></a><span class="in">```{r lda_hipotesis_homogen}</span></span>
<span id="cb49-211"><a href="#cb49-211" aria-hidden="true" tabindex="-1"></a><span class="in">plot_data &lt;- iris_data %&gt;% </span></span>
<span id="cb49-212"><a href="#cb49-212" aria-hidden="true" tabindex="-1"></a><span class="in">  pivot_longer(-Species)</span></span>
<span id="cb49-213"><a href="#cb49-213" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-214"><a href="#cb49-214" aria-hidden="true" tabindex="-1"></a><span class="in">ggplot(plot_data) +</span></span>
<span id="cb49-215"><a href="#cb49-215" aria-hidden="true" tabindex="-1"></a><span class="in">       geom_boxplot(aes(x=name, y=value, fill=Species)) +</span></span>
<span id="cb49-216"><a href="#cb49-216" aria-hidden="true" tabindex="-1"></a><span class="in">       labs(title = "Boxplots for each predictor and target value") +</span></span>
<span id="cb49-217"><a href="#cb49-217" aria-hidden="true" tabindex="-1"></a><span class="in">       scale_fill_ipsum(name ="Species") +</span></span>
<span id="cb49-218"><a href="#cb49-218" aria-hidden="true" tabindex="-1"></a><span class="in">       theme_ipsum_rc()</span></span>
<span id="cb49-219"><a href="#cb49-219" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-220"><a href="#cb49-220" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-221"><a href="#cb49-221" aria-hidden="true" tabindex="-1"></a>Cuantitativamente, podemos calcular las matrices de covarianza o aplicar el test de Bartlett con <span class="co">[</span><span class="ot">`bartlett.test`</span><span class="co">](https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/bartlett.test)</span>.</span>
<span id="cb49-222"><a href="#cb49-222" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-223"><a href="#cb49-223" aria-hidden="true" tabindex="-1"></a><span class="in">```{r lda_bartlett}</span></span>
<span id="cb49-224"><a href="#cb49-224" aria-hidden="true" tabindex="-1"></a><span class="in">iris_data %&gt;%</span></span>
<span id="cb49-225"><a href="#cb49-225" aria-hidden="true" tabindex="-1"></a><span class="in">  group_by(Species) %&gt;%</span></span>
<span id="cb49-226"><a href="#cb49-226" aria-hidden="true" tabindex="-1"></a><span class="in">  group_map(~var(.x))</span></span>
<span id="cb49-227"><a href="#cb49-227" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-228"><a href="#cb49-228" aria-hidden="true" tabindex="-1"></a><span class="in">bartlett.test(Sepal.Length ~ Species, iris_data)</span></span>
<span id="cb49-229"><a href="#cb49-229" aria-hidden="true" tabindex="-1"></a><span class="in">bartlett.test(Sepal.Width  ~ Species, iris_data)</span></span>
<span id="cb49-230"><a href="#cb49-230" aria-hidden="true" tabindex="-1"></a><span class="in">bartlett.test(Petal.Length ~ Species, iris_data)</span></span>
<span id="cb49-231"><a href="#cb49-231" aria-hidden="true" tabindex="-1"></a><span class="in">bartlett.test(Petal.Width  ~ Species, iris_data)</span></span>
<span id="cb49-232"><a href="#cb49-232" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-233"><a href="#cb49-233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-234"><a href="#cb49-234" aria-hidden="true" tabindex="-1"></a><span class="fu">## Particionamiento de datos</span></span>
<span id="cb49-235"><a href="#cb49-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-236"><a href="#cb49-236" aria-hidden="true" tabindex="-1"></a><span class="in">```{r particionamiento_lda}</span></span>
<span id="cb49-237"><a href="#cb49-237" aria-hidden="true" tabindex="-1"></a><span class="in">set.seed(0)</span></span>
<span id="cb49-238"><a href="#cb49-238" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-239"><a href="#cb49-239" aria-hidden="true" tabindex="-1"></a><span class="in">trainIndex &lt;- createDataPartition(iris_scaled$Species, p = .50, list = FALSE)</span></span>
<span id="cb49-240"><a href="#cb49-240" aria-hidden="true" tabindex="-1"></a><span class="in">train &lt;- iris_data[trainIndex, ] </span></span>
<span id="cb49-241"><a href="#cb49-241" aria-hidden="true" tabindex="-1"></a><span class="in">val   &lt;- iris_data[-trainIndex, ]</span></span>
<span id="cb49-242"><a href="#cb49-242" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-243"><a href="#cb49-243" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-244"><a href="#cb49-244" aria-hidden="true" tabindex="-1"></a><span class="fu">## Predicción con LDA</span></span>
<span id="cb49-245"><a href="#cb49-245" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-246"><a href="#cb49-246" aria-hidden="true" tabindex="-1"></a>Podemos crear el modelo LDA para clasificar los datos de validación a partir de las instancias de entrenamiento.</span>
<span id="cb49-247"><a href="#cb49-247" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-248"><a href="#cb49-248" aria-hidden="true" tabindex="-1"></a><span class="in">```{r lda_modelo}</span></span>
<span id="cb49-249"><a href="#cb49-249" aria-hidden="true" tabindex="-1"></a><span class="in">model.lda &lt;- lda(Species ~ Sepal.Length + Sepal.Width + Petal.Length + Petal.Width, data = train)</span></span>
<span id="cb49-250"><a href="#cb49-250" aria-hidden="true" tabindex="-1"></a><span class="in">model.lda</span></span>
<span id="cb49-251"><a href="#cb49-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-252"><a href="#cb49-252" aria-hidden="true" tabindex="-1"></a><span class="in">lda.pred.train &lt;- predict(model.lda, train[1:4])</span></span>
<span id="cb49-253"><a href="#cb49-253" aria-hidden="true" tabindex="-1"></a><span class="in">lda.pred.val &lt;- predict(model.lda, val[1:4])</span></span>
<span id="cb49-254"><a href="#cb49-254" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-255"><a href="#cb49-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-256"><a href="#cb49-256" aria-hidden="true" tabindex="-1"></a>Observar la transformación de los datos usados para crear el modelo.</span>
<span id="cb49-257"><a href="#cb49-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-258"><a href="#cb49-258" aria-hidden="true" tabindex="-1"></a><span class="in">```{r lda_transformacion}</span></span>
<span id="cb49-259"><a href="#cb49-259" aria-hidden="true" tabindex="-1"></a><span class="in">plot_data &lt;- lda.pred.train$x %&gt;%</span></span>
<span id="cb49-260"><a href="#cb49-260" aria-hidden="true" tabindex="-1"></a><span class="in">  as_tibble() %&gt;%</span></span>
<span id="cb49-261"><a href="#cb49-261" aria-hidden="true" tabindex="-1"></a><span class="in">  mutate(Species = train[[5]])</span></span>
<span id="cb49-262"><a href="#cb49-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-263"><a href="#cb49-263" aria-hidden="true" tabindex="-1"></a><span class="in">ggplot(data = plot_data) + </span></span>
<span id="cb49-264"><a href="#cb49-264" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_point(aes(x = LD1, y = LD2, color = Species))  +</span></span>
<span id="cb49-265"><a href="#cb49-265" aria-hidden="true" tabindex="-1"></a><span class="in">  scale_colour_manual(name = "Species", values=c("red", "green", "blue"), labels=c("Iris Setosa", "Iris Versicolor", "Iris Virginica")) %&gt;%</span></span>
<span id="cb49-266"><a href="#cb49-266" aria-hidden="true" tabindex="-1"></a><span class="in">  labs(title = "Data transformed after LDA") +</span></span>
<span id="cb49-267"><a href="#cb49-267" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_ipsum_rc()</span></span>
<span id="cb49-268"><a href="#cb49-268" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-269"><a href="#cb49-269" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-270"><a href="#cb49-270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-271"><a href="#cb49-271" aria-hidden="true" tabindex="-1"></a><span class="fu">## Analisis y validación</span></span>
<span id="cb49-272"><a href="#cb49-272" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-273"><a href="#cb49-273" aria-hidden="true" tabindex="-1"></a>Finalmente, podemos calcular las métricas en validación.</span>
<span id="cb49-274"><a href="#cb49-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-275"><a href="#cb49-275" aria-hidden="true" tabindex="-1"></a><span class="in">```{r lda_metricas}</span></span>
<span id="cb49-276"><a href="#cb49-276" aria-hidden="true" tabindex="-1"></a><span class="in">(t &lt;- table(lda.pred.val$class, val$Species))</span></span>
<span id="cb49-277"><a href="#cb49-277" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-278"><a href="#cb49-278" aria-hidden="true" tabindex="-1"></a><span class="in">val_acc_rate &lt;- sum(diag(t)) / nrow(val)</span></span>
<span id="cb49-279"><a href="#cb49-279" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-280"><a href="#cb49-280" aria-hidden="true" tabindex="-1"></a><span class="in">print(paste0("% de acierto en validación: ", val_acc_rate))</span></span>
<span id="cb49-281"><a href="#cb49-281" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-282"><a href="#cb49-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-283"><a href="#cb49-283" aria-hidden="true" tabindex="-1"></a>Y visualizar las predicciones, tanto con los datos de entrenamiento como los de validación.</span>
<span id="cb49-284"><a href="#cb49-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-285"><a href="#cb49-285" aria-hidden="true" tabindex="-1"></a><span class="in">```{r lda_prediccion_train}</span></span>
<span id="cb49-286"><a href="#cb49-286" aria-hidden="true" tabindex="-1"></a><span class="in">plot_data &lt;- lda.pred.train$x %&gt;%</span></span>
<span id="cb49-287"><a href="#cb49-287" aria-hidden="true" tabindex="-1"></a><span class="in">  as_tibble() %&gt;%</span></span>
<span id="cb49-288"><a href="#cb49-288" aria-hidden="true" tabindex="-1"></a><span class="in">  mutate(known = train[[5]]) %&gt;%</span></span>
<span id="cb49-289"><a href="#cb49-289" aria-hidden="true" tabindex="-1"></a><span class="in">  mutate(prediction = lda.pred.train$class) %&gt;%</span></span>
<span id="cb49-290"><a href="#cb49-290" aria-hidden="true" tabindex="-1"></a><span class="in">  pivot_longer(c("prediction", "known"))</span></span>
<span id="cb49-291"><a href="#cb49-291" aria-hidden="true" tabindex="-1"></a><span class="in">  </span></span>
<span id="cb49-292"><a href="#cb49-292" aria-hidden="true" tabindex="-1"></a><span class="in">ggplot(data = plot_data) + </span></span>
<span id="cb49-293"><a href="#cb49-293" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_point(aes(x = LD1, y = LD2, shape = name, color = value))  +</span></span>
<span id="cb49-294"><a href="#cb49-294" aria-hidden="true" tabindex="-1"></a><span class="in">  scale_colour_manual(name = "Species", values=c("red", "green", "blue"), labels=c("Iris Setosa", "Iris Versicolor", "Iris Virginica")) + </span></span>
<span id="cb49-295"><a href="#cb49-295" aria-hidden="true" tabindex="-1"></a><span class="in">  scale_shape_manual(name = "Type", values=c(5, 3)) +</span></span>
<span id="cb49-296"><a href="#cb49-296" aria-hidden="true" tabindex="-1"></a><span class="in">  labs(title = "Train data + predictions transformed after LDA") +</span></span>
<span id="cb49-297"><a href="#cb49-297" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_ipsum_rc()</span></span>
<span id="cb49-298"><a href="#cb49-298" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-299"><a href="#cb49-299" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-300"><a href="#cb49-300" aria-hidden="true" tabindex="-1"></a><span class="in">```{r lda_prediccion_val}</span></span>
<span id="cb49-301"><a href="#cb49-301" aria-hidden="true" tabindex="-1"></a><span class="in">plot_data &lt;- lda.pred.val$x %&gt;%</span></span>
<span id="cb49-302"><a href="#cb49-302" aria-hidden="true" tabindex="-1"></a><span class="in">  as_tibble() %&gt;%</span></span>
<span id="cb49-303"><a href="#cb49-303" aria-hidden="true" tabindex="-1"></a><span class="in">  mutate(known = val[[5]]) %&gt;%</span></span>
<span id="cb49-304"><a href="#cb49-304" aria-hidden="true" tabindex="-1"></a><span class="in">  mutate(prediction = lda.pred.val$class) %&gt;%</span></span>
<span id="cb49-305"><a href="#cb49-305" aria-hidden="true" tabindex="-1"></a><span class="in">  pivot_longer(c("prediction", "known"))</span></span>
<span id="cb49-306"><a href="#cb49-306" aria-hidden="true" tabindex="-1"></a><span class="in">  </span></span>
<span id="cb49-307"><a href="#cb49-307" aria-hidden="true" tabindex="-1"></a><span class="in">ggplot(data = plot_data) + </span></span>
<span id="cb49-308"><a href="#cb49-308" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_point(aes(x = LD1, y = LD2, shape = name, color = value))  +</span></span>
<span id="cb49-309"><a href="#cb49-309" aria-hidden="true" tabindex="-1"></a><span class="in">  scale_colour_manual(name = "Species", values=c("red", "green", "blue"), labels=c("Iris Setosa", "Iris Versicolor", "Iris Virginica")) + </span></span>
<span id="cb49-310"><a href="#cb49-310" aria-hidden="true" tabindex="-1"></a><span class="in">  scale_shape_manual(name = "Type", values=c(5, 3)) +</span></span>
<span id="cb49-311"><a href="#cb49-311" aria-hidden="true" tabindex="-1"></a><span class="in">  labs(title = "Validation data + predictions transformed after LDA") +</span></span>
<span id="cb49-312"><a href="#cb49-312" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_ipsum_rc()</span></span>
<span id="cb49-313"><a href="#cb49-313" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-314"><a href="#cb49-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-315"><a href="#cb49-315" aria-hidden="true" tabindex="-1"></a><span class="fu"># Clasificación con QDA</span></span>
<span id="cb49-316"><a href="#cb49-316" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-317"><a href="#cb49-317" aria-hidden="true" tabindex="-1"></a>Por último, creamos un clasificador con QDA usando el método <span class="co">[</span><span class="ot">`qda`</span><span class="co">](https://www.rdocumentation.org/packages/MASS/versions/7.3-58.3/topics/qda)</span> de <span class="co">[</span><span class="ot">`MASS`</span><span class="co">](https://www.rdocumentation.org/packages/MASS/versions/7.3-58.3)</span>.</span>
<span id="cb49-318"><a href="#cb49-318" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-319"><a href="#cb49-319" aria-hidden="true" tabindex="-1"></a><span class="in">```{r qda}</span></span>
<span id="cb49-320"><a href="#cb49-320" aria-hidden="true" tabindex="-1"></a><span class="in">model.qda &lt;- qda(Species ~ Sepal.Length + Sepal.Width + Petal.Length + Petal.Width, data = train)</span></span>
<span id="cb49-321"><a href="#cb49-321" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-322"><a href="#cb49-322" aria-hidden="true" tabindex="-1"></a><span class="in">qda.pred.train &lt;- predict(model.qda, train[1:4])</span></span>
<span id="cb49-323"><a href="#cb49-323" aria-hidden="true" tabindex="-1"></a><span class="in">qda.pred.val &lt;- predict(model.qda, val[1:4])</span></span>
<span id="cb49-324"><a href="#cb49-324" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-325"><a href="#cb49-325" aria-hidden="true" tabindex="-1"></a><span class="in">(t &lt;- table(qda.pred.val$class, val$Species))</span></span>
<span id="cb49-326"><a href="#cb49-326" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-327"><a href="#cb49-327" aria-hidden="true" tabindex="-1"></a><span class="in">val_acc_rate &lt;- sum(diag(t)) / nrow(val)</span></span>
<span id="cb49-328"><a href="#cb49-328" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-329"><a href="#cb49-329" aria-hidden="true" tabindex="-1"></a><span class="in">print(paste0("% de acierto en validación: ", val_acc_rate))</span></span>
<span id="cb49-330"><a href="#cb49-330" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb49-331"><a href="#cb49-331" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-332"><a href="#cb49-332" aria-hidden="true" tabindex="-1"></a><span class="fu"># LDA y QDA en `caret`</span></span>
<span id="cb49-333"><a href="#cb49-333" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-334"><a href="#cb49-334" aria-hidden="true" tabindex="-1"></a><span class="in">```{r lda_qda_caret}</span></span>
<span id="cb49-335"><a href="#cb49-335" aria-hidden="true" tabindex="-1"></a><span class="in">suppressWarnings({</span></span>
<span id="cb49-336"><a href="#cb49-336" aria-hidden="true" tabindex="-1"></a><span class="in">  lda.model &lt;- train(train[1:4],</span></span>
<span id="cb49-337"><a href="#cb49-337" aria-hidden="true" tabindex="-1"></a><span class="in">                    train$Species,</span></span>
<span id="cb49-338"><a href="#cb49-338" aria-hidden="true" tabindex="-1"></a><span class="in">                    method="lda", </span></span>
<span id="cb49-339"><a href="#cb49-339" aria-hidden="true" tabindex="-1"></a><span class="in">                    preProcess = c("center", "scale"),</span></span>
<span id="cb49-340"><a href="#cb49-340" aria-hidden="true" tabindex="-1"></a><span class="in">                    tuneLength = 10,</span></span>
<span id="cb49-341"><a href="#cb49-341" aria-hidden="true" tabindex="-1"></a><span class="in">                    trControl = trainControl(method = "cv"))</span></span>
<span id="cb49-342"><a href="#cb49-342" aria-hidden="true" tabindex="-1"></a><span class="in">})</span></span>
<span id="cb49-343"><a href="#cb49-343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-344"><a href="#cb49-344" aria-hidden="true" tabindex="-1"></a><span class="in">lda.model</span></span>
<span id="cb49-345"><a href="#cb49-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-346"><a href="#cb49-346" aria-hidden="true" tabindex="-1"></a><span class="in">suppressWarnings({</span></span>
<span id="cb49-347"><a href="#cb49-347" aria-hidden="true" tabindex="-1"></a><span class="in">  qda.model &lt;- train(train[1:4],</span></span>
<span id="cb49-348"><a href="#cb49-348" aria-hidden="true" tabindex="-1"></a><span class="in">                    train$Species,</span></span>
<span id="cb49-349"><a href="#cb49-349" aria-hidden="true" tabindex="-1"></a><span class="in">                    method="qda", </span></span>
<span id="cb49-350"><a href="#cb49-350" aria-hidden="true" tabindex="-1"></a><span class="in">                    preProcess = c("center", "scale"),</span></span>
<span id="cb49-351"><a href="#cb49-351" aria-hidden="true" tabindex="-1"></a><span class="in">                    tuneLength = 10,</span></span>
<span id="cb49-352"><a href="#cb49-352" aria-hidden="true" tabindex="-1"></a><span class="in">                    trControl = trainControl(method = "cv"))</span></span>
<span id="cb49-353"><a href="#cb49-353" aria-hidden="true" tabindex="-1"></a><span class="in">})</span></span>
<span id="cb49-354"><a href="#cb49-354" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-355"><a href="#cb49-355" aria-hidden="true" tabindex="-1"></a><span class="in">qda.model</span></span>
<span id="cb49-356"><a href="#cb49-356" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span></code><button title="Copiar al portapapeles" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->




</body></html>